src/nccl_ofi_rdma.c

```diff
@@ -5503,9 +5468,24 @@ static int connect(nccl_net_ofi_ep_t *base_ep,
 
 		comm_state->stage = COMM_CONN_RESP_REQ_PENDING;
 
+	case COMM_CONN_RESP_REQ_PENDING:
+
+		// TODO: We might want to use READ_ONCE to read variable `connected'
+		if (!s_comm->connected) {
+			ret = ofi_process_cq(ep);
+
+			/*
+			 * To allow `connect` side to continue to process its cq,
+			 * don't return a valid send communicator to NCCL until
+			 * the connection is finalized.
+			 */
+			return ret;
+		}
+
+		comm_state->stage = COMM_CONNECTED;
+
 		break;
```

Comment:
Nit: I think it makes more sense to have this `break` after `case COMM_CONNECTED` instead of here, because that is when the state machine is complete. I don't expect that the next 2 `case` statements are matched ever, but if we have them here, it should be written correctly. 
That would also mean that if NCCL calls `connect` after the connection was established, it would return successfully (from what I know the APIs don't say what should be the right behavior in this case).

Reply:
I think it was written this way intentionally to catch the case where NCCL calls `connect` on a handle that has already completed connection establishment. If we return successfully on the second `connect` call, we will have returned two identical send_comm's to NCCL, which will cause errors later. So to me it seems better to catch the case here and error out.